src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/FindPal.cs

```diff
@@ -0,0 +1,268 @@
+                        break;
+                    }
+                    case X509FindType.FindByTimeExpired:
+                    {
+                        DateTime dateTime = ConfirmedCast<DateTime>(findValue);
+                        findPal.FindByTimeExpired(dateTime);
+                        break;
+                    }
+                    case X509FindType.FindByTemplateName:
+                    {
+                        string expected = ConfirmedCast<string>(findValue);
+                        findPal.FindByTemplateName(expected);
+                        break;
+                    }
+                    case X509FindType.FindByApplicationPolicy:
+                    {
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.Policy);
+                        findPal.FindByApplicationPolicy(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByCertificatePolicy:
+                    {
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.Policy);
+                        findPal.FindByCertificatePolicy(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByExtension:
+                    {
+                        string oidValue = ConfirmedOidValue(findPal, findValue, OidGroup.ExtensionOrAttribute);
+                        findPal.FindByExtension(oidValue);
+                        break;
+                    }
+                    case X509FindType.FindByKeyUsage:
+                    {
+                        X509KeyUsageFlags keyUsage = ConfirmedX509KeyUsage(findValue);
+                        findPal.FindByKeyUsage(keyUsage);
+                        break;
+                    }
+                    case X509FindType.FindBySubjectKeyIdentifier:
+                    {
+                        byte[] keyIdentifier = ConfirmedCast<string>(findValue).DecodeHexString();
+                        findPal.FindBySubjectKeyIdentifier(keyIdentifier);
+                        break;
+                    }
+                    default:
+                        throw new CryptographicException(SR.Cryptography_X509_InvalidFindType);
+                }
+            }
+           
+            return results;
+        }
+
+        private static T ConfirmedCast<T>(object findValue)
+        {
+            Debug.Assert(findValue != null);
+
+            if (findValue.GetType() != typeof(T))
+                throw new CryptographicException(SR.Cryptography_X509_InvalidFindValue);
+
+            return (T)findValue;
+        }
+
+        private static string ConfirmedOidValue(IFindPal findPal, object findValue, OidGroup oidGroup)
+        {
+            string maybeOid = ConfirmedCast<string>(findValue);
+
+            if (maybeOid.Length == 0)
+            {
+                throw new ArgumentException(SR.Argument_InvalidOidValue);
+            }
```

Comment:
Just verifying that ArgumentException is the right exception type for this situation?  Is the OID actually coming in through some public API, and it's the argument to that method we're saying is invalid?